sds_data_manager/lambda_code/SDSCode/pointing_frame_handler.py

```diff
@@ -0,0 +1,248 @@
+
+    Returns
+    -------
+    et_start : float
+        Start time of ck_kernel.
+    et_end : float
+        End time of ck_kernel.
+    et_times : numpy.ndarray
+        Array of times between et_start and et_end.
+    """
+    # Get the spacecraft ID.
+    # https://spiceypy.readthedocs.io/en/main/documentation.html#spiceypy.spiceypy.gipool
+    id_imap_spacecraft = spice.gipool("FRAME_IMAP_SPACECRAFT", 0, 1)
+
+    # TODO: Queried pointing start and stop times here.
+
+    # Get the coverage window
+    # https://spiceypy.readthedocs.io/en/main/documentation.html#spiceypy.spiceypy.ckcov
+    cover = spice.ckcov(ck_kernel, int(id_imap_spacecraft), True, "SEGMENT", 0, "TDB")
+    # https://spiceypy.readthedocs.io/en/main/documentation.html#spiceypy.spiceypy.wnfetd
+    et_start, et_end = spice.wnfetd(cover, 0)
+    et_times = np.arange(et_start, et_end, STEP)
+
+    return et_start, et_end, et_times
+
+
+def average_quaternions(et_times):
+    """Average the quaternions.
+
+    Parameters
+    ----------
+    et_times : numpy.ndarray
+        Array of times between et_start and et_end.
+
+    Returns
+    -------
+    q_avg : np.array
+        Average quaternion.
+    z_eclip_time : list
+        Z-axis of the ECLIPJ2000 frame. Used for plotting.
+    """
+    body_quats = []
+    z_eclip_time = []
+    aggregate = np.zeros((4, 4))
+
+    for tdb in et_times:
+        # Rotation matrix from IMAP spacecraft frame to ECLIPJ2000.
+        # https://spiceypy.readthedocs.io/en/main/documentation.html#spiceypy.spiceypy.pxform
+        body_rots = spice.pxform("IMAP_SPACECRAFT", "ECLIPJ2000", tdb)
+        # Convert rotation matrix to quaternion.
+        # https://spiceypy.readthedocs.io/en/main/documentation.html#spiceypy.spiceypy.m2q
+        body_quat = spice.m2q(body_rots)
+        body_quats.append(body_quat)
+        # z-axis of the ECLIPJ2000 frame.
+        z_eclip_time.append(body_rots[:, 2])
+
+        # Standardize the quaternion so that they may be compared.
+        if body_quat[0] < 0:
+            body_quat = -body_quat
+
+        # Aggregate quaternions into a single matrix.
+        aggregate += np.outer(body_quat, body_quat)
+
+    # Reference: Claus Gramkow "On Averaging Rotations"
+    # Link: https://link.springer.com/content/pdf/10.1023/A:1011129215388.pdf
+    aggregate /= len(et_times)
+
+    # Compute eigen values and vectors of the matrix A
+    # Eigenvalues tell you how much "influence" each
+    # direction (eigenvector) has.
+    # The largest eigenvalue corresponds to the direction
+    # that has the most influence.
+    # The eigenvector corresponding to the largest
+    # eigenvalue points in the direction that has the most
+    # combined rotation influence.
+    eigvals, eigvecs = np.linalg.eig(aggregate)
+    # q0: The scalar part of the quaternion.
+    # q1, q2, q3: The vector part of the quaternion.
+    q_avg = eigvecs[:, np.argmax(eigvals)]
+
+    return q_avg, z_eclip_time
```

Comment:
Unused in the calling function?
```suggestion
    return q_avg
```

Reply:
This was used in the tests.